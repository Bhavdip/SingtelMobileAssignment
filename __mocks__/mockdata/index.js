const mockShuffledData = [10, 12, 13, 14, 15, 4, 12, 13, 14, 15, 10];
const mockResolvedData = [
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false,
  false
];
export { mockShuffledData, mockResolvedData };
