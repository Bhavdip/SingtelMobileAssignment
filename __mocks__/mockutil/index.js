import { Platform, I18nManager } from 'react-native';

export const findComponentByID = (shallowWrapperCompt, findComponent, componentTestID) => {
  const foundComponent = shallowWrapperCompt.find(findComponent).findWhere(element => {
    if (element.props().testID === componentTestID) {
      // console.log(element.props());
      return element;
    }
    return null;
  });
  return foundComponent;
};

// Ref: https://stackoverflow.com/questions/43161416/mocking-platform-detection-in-jest-and-react-native
// https://www.codepile.net/pile/Yr5DkK4n
export const setPlatform = platform => {
  if (platform === 'android' || platform === 'ios') {
    Object.defineProperty(Platform, 'OS', {
      get: jest.fn(() => platform)
    });
  }
};

// https://stackoverflow.com/questions/40449434/mocking-globals-in-jest
export const setI18nManager = isRTLValue => {
  Object.defineProperty(I18nManager, 'isRTL', { value: isRTLValue, writable: true });
};
