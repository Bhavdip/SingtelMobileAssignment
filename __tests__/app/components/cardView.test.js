import React, { useState, useRef } from 'react';
import { shallow } from 'enzyme';
import { TouchableOpacity } from 'react-native';
import CardView from '../../../app/components/cardview';
import { findComponentByID } from '../../../__mocks__/mockutil';
import CardFlip from '../../../app/components/cardflip';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
  useRef: jest.fn()
}));


const setUpComponent = (props = {}) => {
  return shallow(<CardView {...props} />);
};

const props = {
  cardId: 10,
  cardNumber: 98,
  isBlocked: false,
  isResolved: false,
  onPressCard: jest.fn(),
  onCardFlipEnd: jest.fn()
};

describe('Test CardView Component | onFlipEnd Function Invoke', () => {
  let wrapperComponent = {};
  let setFrontDisplay;
  beforeEach(() => {
    jest.resetAllMocks();
    setFrontDisplay = jest.fn();
    const fontDisplayMock = (frontDisplay) => [frontDisplay, setFrontDisplay];
    jest.spyOn(React, 'useState').mockImplementation(fontDisplayMock);
    wrapperComponent = setUpComponent(props);
  });
  it('Test Snapshot Match', () => {
    expect(wrapperComponent).toMatchSnapshot();
  });
  it('Test CardFlip onFlipEnd should invoke onCardFlipEnd and font display make it true', () => {
    const CardFlipComp = findComponentByID(wrapperComponent, CardFlip, 'CardFlipComp');
    expect(CardFlipComp.props().testID).toEqual('CardFlipComp');
    CardFlipComp.props().onFlipEnd(0);
    expect(setFrontDisplay).toHaveBeenCalledTimes(1);
    CardFlipComp.props().onFlipEnd(1);
    expect(setFrontDisplay).toHaveBeenCalledTimes(2);
  });
});

describe('Test CardView onCardPress method',() =>{
  let wrapperComponent;
  // on Flip Mock
  const mockFlip = jest.fn();
  let mUseRefCardFlip;
  beforeEach(() => {
    jest.resetAllMocks();
    // set default value as true
    const setFlipPressed = jest.fn()
    jest.spyOn(React, 'useState').mockImplementation(isFlipPress => [isFlipPress=true, setFlipPressed]);
    // Mock useReference
    const mRef = { current: { flip: mockFlip } };
    mUseRefCardFlip = jest.spyOn(React, 'useRef').mockReturnValueOnce(mRef);
    wrapperComponent = setUpComponent(props);
  })
  it('Test CardFlip onCardPress when card is not resolved and it is not blocked', () => {
    const CardFrontSide = findComponentByID(wrapperComponent, TouchableOpacity, 'FrontSide');
    expect(CardFrontSide.props().testID).toEqual('FrontSide');
    CardFrontSide.props().onPress();
    expect(mUseRefCardFlip).toHaveBeenCalled();
  });
  it('Test CardFlip onCardPress when card is either resolved or blocked', () => {
    wrapperComponent.setProps({ onPressCard: null, isResolved: true, isBlocked: true });
    const CardFrontSide = findComponentByID(wrapperComponent, TouchableOpacity, 'FrontSide');
    expect(CardFrontSide.props().testID).toEqual('FrontSide');
    CardFrontSide.props().onPress();
    expect(mUseRefCardFlip).toHaveBeenCalled();
  });
  it('Test CardFlip onCardFlipEnd function', () => {
    const CardFlipComp = findComponentByID(wrapperComponent, CardFlip, 'CardFlipComp');
    expect(CardFlipComp.props().testID).toEqual('CardFlipComp');
    CardFlipComp.props().onFlipEnd(0);
  });
})

