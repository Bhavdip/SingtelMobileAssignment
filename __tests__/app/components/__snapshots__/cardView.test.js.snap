// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CardView Component | onFlipEnd Function Invoke Test Snapshot Match 1`] = `ShallowWrapper {}`;
